node/subsystem/src/util.rs

```diff
@@ -499,9 +499,9 @@ where
 				}
 			}
 			Ok(Signal(Conclude)) => {
-				// breaking the loop ends fn run, which drops `jobs`, which immediately drops all ongoing work
-				//
-				// REVIEW: is this the behavior we want?
+				// Breaking the loop ends fn run, which drops `jobs`, which immediately drops all ongoing work.
+				// We can afford to wait a little while to shut them all down properly before doing that.
+				future::join_all(jobs.running.drain().map(|(_, handle)| handle.stop())).await;
```

Comment:
```suggestion
let future = jobs.drain().collect::<FuturesUnordered<_>>().into_future();
future.await;
```
I read that this is more efficient in polling than using `join_all` which polls all of them always (mixed that up, but Pierre actually smartened me up).

Reply:
After about an hour of attacking this problem, I ended up with this implementation:

```rust
// Forwarding the stream to a drain means we wait until all of the items in the stream
// have completed. Contrast with `into_future`, which turns it into a future of `(head, rest_stream)`.
jobs.running
	.drain()
	.map(|(_, handle)| handle.stop())
	.collect::<FuturesUnordered<_>>()
	.forward(drain())
	.await;
```

Which produces these errors:

```
    Checking polkadot-node-subsystem v0.1.0 (/home/coriolinus/Documents/Projects/paritytech/polkadot/node/subsystem)
error[E0271]: type mismatch resolving `<futures_util::stream::futures_unordered::FuturesUnordered<impl core::future::future::Future> as futures_core::stream::Stream>::Item == std::result::Result<_, _>`
   --> node/subsystem/src/util.rs:563:7
    |
563 |                     .forward(drain())
    |                      ^^^^^^^ expected `()`, found enum `std::result::Result`
    |
    = note: expected unit type `()`
                    found enum `std::result::Result<_, _>`
    = note: required because of the requirements on the impl of `futures_core::stream::TryStream` for `futures_util::stream::futures_unordered::FuturesUnordered<impl core::future::future::Future>`

error[E0271]: type mismatch resolving `<futures_util::stream::futures_unordered::FuturesUnordered<impl core::future::future::Future> as futures_core::stream::Stream>::Item == std::result::Result<_, std::convert::Infallible>`
   --> node/subsystem/src/util.rs:559:51
    |
559 |                   let _: Result<(), std::convert::Infallible> = jobs.running
    |  _______________________________________________________________^
560 | |                     .drain()
561 | |                     .map(|(_, handle)| handle.stop())
562 | |                     .collect::<FuturesUnordered<_>>()
563 | |                     .forward(drain())
564 | |                     .await;
    | |__________________________^ expected `()`, found enum `std::result::Result`
    |
    = note: expected unit type `()`
                    found enum `std::result::Result<_, std::convert::Infallible>`
    = note: required because of the requirements on the impl of `core::future::future::Future` for `futures_util::stream::stream::forward::Forward<futures_util::stream::futures_unordered::FuturesUnordered<impl core::future::future::Future>, futures_util::sink::drain::Drain<_>, _>`

error[E0271]: type mismatch resolving `<futures_util::stream::futures_unordered::FuturesUnordered<impl core::future::future::Future> as futures_core::stream::Stream>::Item == std::result::Result<_, _>`
   --> node/subsystem/src/util.rs:559:51
    |
559 |                   let _: Result<(), std::convert::Infallible> = jobs.running
    |  _______________________________________________________________^
560 | |                     .drain()
561 | |                     .map(|(_, handle)| handle.stop())
562 | |                     .collect::<FuturesUnordered<_>>()
563 | |                     .forward(drain())
564 | |                     .await;
    | |__________________________^ expected `()`, found enum `std::result::Result`
    |
    = note: expected unit type `()`
                    found enum `std::result::Result<_, _>`
    = note: required because of the requirements on the impl of `futures_core::stream::TryStream` for `futures_util::stream::futures_unordered::FuturesUnordered<impl core::future::future::Future>`
    = note: required because of the requirements on the impl of `core::future::future::Future` for `futures_util::stream::stream::Forward<futures_util::stream::futures_unordered::FuturesUnordered<impl core::future::future::Future>, futures_util::sink::drain::Drain<_>>`

error: aborting due to 3 previous errors
```

I get that somewhere in that chain, it's confusing `()` and `Result<(), std::convert::Infallible>`, but I have no idea how to actually fix the problem; nothing I've attempted has solved it. If you know the way to solve those errors, I'd be glad to hear it!

I like the idea of this suggestion in principle, but for now I'm going to give up on it. I suspect that we'll have at most a few dozen jobs running at any given moment, so the inefficient implementation is probably good enough for practical purposes.

Reply:
```suggestion
				use futures::stream::StreamExt;
				use futures::stream::TryStreamExt;
				use futures::stream::FuturesUnordered;
				let unordered = jobs.running
					.drain()
					.map(|(_, handle)| handle.stop())
					.collect::<FuturesUnordered<_>>();
				let _ = async move { unordered.collect::<Vec<_>>() } .await;

```

compiles locally :)

Reply:
:+1: